homeassistant/components/nest/config_flow.py

```diff
@@ -243,25 +280,117 @@ async def async_step_reauth_confirm(
             entry = next(iter(existing_entries))
             if "auth_implementation" in entry.data:
                 data = {"implementation": entry.data["auth_implementation"]}
-                return await self.async_step_user(data)
+                return await super().async_step_user(data)
         return await self.async_step_user()
 
     async def async_step_user(
         self, user_input: dict[str, Any] | None = None
     ) -> FlowResult:
         """Handle a flow initialized by the user."""
+        if self.config_mode == ConfigMode.LEGACY:
+            return await self.async_step_init(user_input)
+        self._data[DATA_SDM] = {}
+        # Reauth will update an existing entry
+        entries = self._async_current_entries()
+        if entries and not self._reauth:
+            return self.async_abort(reason="single_instance_allowed")
         if self.config_mode == ConfigMode.SDM:
-            # Reauth will update an existing entry
-            if self._async_current_entries() and not self._reauth:
-                return self.async_abort(reason="single_instance_allowed")
             return await super().async_step_user(user_input)
-        return await self.async_step_init(user_input)
+        # ConfigMode.SDM_CONFIG_ENTRY registers OAuth impl during the config
+        # and can be changed if the flow is restarted to fix a mistake.
+        return await self.async_step_cloud_project()
+
+    async def async_step_cloud_project(
+        self, user_input: dict | None = None
+    ) -> FlowResult:
+        """Handle cloud project in user input."""
+        if user_input is not None:
+            self._data.update(user_input)
+            return await self.async_step_oauth_input()
+
+        return self.async_show_form(
+            step_id="cloud_project",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_CLOUD_PROJECT_ID): str,
+                }
+            ),
+            description_placeholders={
+                "cloud_console_url": CLOUD_CONSOLE_URL,
+                "sdm_api_url": SDM_API_URL,
+                "pubsub_api_url": PUBSUB_API_URL,
+                "more_info_url": MORE_INFO_URL_FORMAT.format(
+                    anchor="step_cloud_project"
+                ),
+            },
+        )
+
+    async def async_step_oauth_input(
+        self, user_input: dict | None = None
+    ) -> FlowResult:
+        """Create a new OAuth implementation from user input."""
+        if user_input is not None:
+            self._data.update(user_input)
+            return await self.async_step_device_project()
+        cloud_project_id = self._data[CONF_CLOUD_PROJECT_ID]
+        return self.async_show_form(
+            step_id="oauth_input",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_CLIENT_ID): str,
+                    vol.Required(CONF_CLIENT_SECRET): str,
```

Comment:
We don't allow this currently, and I'm against adding it here as well.

I suggest we do not start working on this further until a consensus has been reached.